[SYCL][HIP] Support of AMD matrix core instructions

mmoadeli opened the pull request:
* Support one block AMD matrix core instructions for `__gfx90a__ architecture.
* Supports `__builtin_amdgcn_mfma_i32_32x32x8i8`, `__builtin_amdgcn_mfma_i32_16x16x16i8`, `__builtin_amdgcn_mfma_f64_16x16x4f64`, `__builtin_amdgcn_mfma_f32_32x32x8bf16_1k`, `__builtin_amdgcn_mfma_f32_16x16x16bf16_1k`, `__builtin_amdgcn_mfma_f32_32x32x8f16` and `__builtin_amdgcn_mfma_f32_16x16x16f16` instructions.

yubingex007-a11y committed:
[Matrix] syntax changes as prepraration before moving joint matrix from
experimental namespace
As part of the effort to move joint matrix from experimental namespace to supported. A review of the API is being done as part of https://github.com/intel/llvm/pull/7964. This results in the following changes in the syntax:
1- Add Td to joint_matrix_mad as Tc can be different from Td on the GPU,
Now, we make D as an input argument to mad.
2-  Change “packed” to ext_intel_packed:
3-  Move EWOps (get_wi_data, wi_element, get_coord) to detail namespace)
4- add const to joint_matrix in store and mad
5 - add joint_matrix_copy/assignment function
6- add apply with coordination (change existing tests)
7- change get_coord vector type from int32_t to size_t
8- delete explicitly both = and copy ctor.

yubingex007-a11y committed:
clang-format

yubingex007-a11y committed:
fix typo: dest->dst

yubingex007-a11y committed:
fix testcase

yubingex007-a11y committed:
fix mad bug

yubingex007-a11y committed:
fix cuda const joint_matrix_cuda

yubingex007-a11y committed:
fix const issue of jm_store_cuda

yubingex007-a11y committed:
fix const

yubingex007-a11y committed:
lint

yubingex007-a11y committed:
address dounia's comments and roll back all the testcase changes

yubingex007-a11y committed:
test changes: mov D in mad

yubingex007-a11y committed:
testcase changes: ext_intel_layout

yubingex007-a11y committed:
testcase changes: wi_data=>jm_apply

yubingex007-a11y committed:
lint

yubingex007-a11y committed:
Merge remote-tracking branch 'intel_llvm/sycl' into jm_syntax

yubingex007-a11y committed:
Merge remote-tracking branch 'intel_llvm/sycl' into jm_syntax

yubingex007-a11y committed:
handle cuda testcase compfail

yubingex007-a11y committed:
address dounia's comments

yubingex007-a11y committed:
lint

yubingex007-a11y committed:
rm sycl/test/matrix/query-use.cpp

yubingex007-a11y committed:
fix x jm_mad in joint_matrix_bf16_fill_k_cache_impl.hpp

Commit:
* Support one block AMD matrix core instructions

* Add HIP matrix core tests.
* Add `gpu-amd-gfx90a` as a feature in lig.cfg.py.

Add HIP matrix core support into joint_matrix documentation.

JackAKirk commented on sycl/include/sycl/ext/oneapi/matrix/matrix-hip.hpp:
```diff
@@ -0,0 +1,317 @@
+  if constexpr (Layout == matrix_layout::row_major) {
+    res.data = {0};
+  } else if constexpr (Layout == matrix_layout::col_major) {
+    res.data = {0};
```
So it seems this doesn't load any data to the accumulator? I guess you'll add this before opening a PR?

JackAKirk commented on sycl/include/sycl/ext/oneapi/matrix/matrix-hip.hpp:
```diff
@@ -0,0 +1,317 @@
+    access::decorated IsDecorated,
+    typename = typename std::enable_if_t<(Layout == matrix_layout::row_major ||
+                                          Layout == matrix_layout::col_major)>>
+void load_multiplicand_hip(
```
Have you tried loading from const matrix? I'm guessing the type checks will break it below?

Might have to do something similar to https://github.com/intel/llvm/pull/6532/files

mmoadeli replied:
The type checks have been on `S` which is the type joint_matrix created from. 
Even if `T`(multi_ptr type) is const, it should be fine, as the `const` value is assigned to non-const value of `joint_matrix` data.

JackAKirk commented on sycl/include/sycl/ext/oneapi/matrix/matrix-hip.hpp:
```diff
@@ -0,0 +1,317 @@
+      const auto thread_x = idx % NumCols;
+      const auto thread_y = idx / NumCols;
+      for (int i = 0; i < 4; ++i) {
+        dst[thread_x + i * 4 * NumCols + thread_y * NumCols] = src.data[i];
```
I wonder if this loop gets unrolled automatically or if not whether it affects perf?

mmoadeli replied:
That's fair point. I'd suggest this to investigated it later.

JackAKirk commented on sycl/include/sycl/ext/oneapi/matrix/matrix-hip.hpp:
```diff
@@ -0,0 +1,317 @@
+      }
+    }
+  } else {
+    static_assert(false && "Invalid dimenstions!");
```
Is it possible to define a joint_matrix that can lead to this assert? Ideally it invalid types/sizes can't be instantiated, such that asserts like these aren't required.

JackAKirk commented on sycl/include/sycl/ext/oneapi/matrix/matrix-hip.hpp:
```diff
@@ -0,0 +1,317 @@
+    store_layoutT<Group, matrix_layout::col_major>(src, dst, stride, sg);
+    break;
+  default:
+    assert(false && "Invalid layout specified!");
```
I think only row_major or col_major layouts can be specified? Maybe I am wrong, but if this is the case such asserts aren't required.

JackAKirk commented on sycl/doc/extensions/experimental/sycl_ext_matrix/sycl_ext_oneapi_matrix.asciidoc:
```diff
@@ -964,6 +964,35 @@ multiple of 4 when `T` is `float`; where `T` is the type of the
+Similarly, C and D matrices must share the same type.
+
+IMPORTANT: Currently, only one block AMD Matrix Core instructions in GFX90A and 
+GFX908 (MI200, MI210, MI250 and MI250X GPUs) architecture are supported.
```
can you specify which are GFX908 and which are GFX90a?

yubingex007-a11y committed:
Merge remote-tracking branch 'intel_llvm/sycl' into jm_syntax

Commit:
* Update matrix core support into joint_matrix documentation.
* Use fully qualified names.
* Add diagnostic to tell the user that joint_maitrx is only supported for gfx90a.
* Remove unnecessary `else` conditions.
* Merge HIP matrix cpp files.

Commit:
- Fix adding `gpu-amd-gfx90a` as available feature.
- Update comment with `GFX90A` info.

Commit:
Fix a missing variable name change.

Commit:
Add decoration type for call to get_multi_ptr function.

yubingex007-a11y committed:
address comments

yubingex007-a11y committed:
Merge remote-tracking branch 'intel_llvm/sycl' into jm_syntax

yubingex007-a11y committed:
rm element_wise_irreg_sum_rows_impl.hpp

yubingex007-a11y committed:
small fix

yubingex007-a11y committed:
small fix

Commit:
Update use cases of `mad` to have variables holding result of `mad` as a parameter of the function.

Commit:
Implement joint_matrix_copy for HIP(gfx90a) backend.

Commit:
std::ignore unused input parameters.

Commit:
std::ignore unused `sg` parameters in`joint_matrix_store`.

Commit:
Merge branch 'sycl' into amd-joint-matrix

Commit:
Fix AMD `joint_matrix_copy` function.

Commit:
- Fix joint_matrix_hip_copy.
- Initialize the matrix with random values.

Commit:
Remove curly braces for initialization of joint_matrix.

Commit:
- Use sycl::marray as container for jont_matrix data.
- Update joint_matrix_apply / test.
- Remove un-necessary code from test.

Commit:
Modify `joint_matrix_apply` test.

Commit:
Update allow difference after using matrix random input.

Commit:
Fix `hip_matrix_copy`.

Commit:
- Improve hip mfma tests to support matrices of multiple of K size.
- Fix 'load / double' to to support matrices of multiple of K size.

Commit:
Fix call to `copy` and `fill` for hip joint matrix.

Commit:
Fix reference mma calculation.

Commit:
Rename `cuda_impl` and `hip_impl` member of `joint_matrix` to `matrix_impl`

Commit:
Merge branch 'sycl' into amd-joint-matrix

Commit:
Replace `cuda_impl` with `matrix_impl`

Commit:
Rename `data` in `joint_matrix_hip` with `wi_marray`.
Use same code for `copy`, `fill` and `apply`.
Remove `-DSYCL_EXT_ONEAPI_MATRIX_VERSION=4`

Commit:
Improve `joint_matrix_copy` by avoiding the loop.

Commit:
Add a missing `comma` to the test.

Commit:
Remove `-DSYCL_EXT_ONEAPI_MATRIX_VERSION=4` from AMD matrix compilation arguments.

Commit:
Guard including `matrix-hip.hpp`

Commit:
Minor macro readability improvement.
